apps/passport-server/src/database/queries/pipelineDefinitionDB.ts

```diff
@@ -234,7 +251,51 @@ export class PipelineDefinitionDB implements IPipelineDefinitionDB {
     definitions: PipelineDefinition[]
   ): Promise<void> {
     for (const definition of definitions) {
-      await this.upsertDefinition(definition);
+      await this.upsertDefinition(definition, undefined);
     }
   }
+
+  public async appendToEditHistory(
+    pipelineDefinition: PipelineDefinition,
+    editorUserId?: string
+  ): Promise<void> {
+    await this.db.query(
+      `
+    insert into podbox_edit_history
+    (pipeline, editor_user_id, time_created)
+    values
+    ($1, $2, $3);`,
+      [pipelineDefinition, editorUserId, new Date()]
+    );
+  }
+
+  public async getEditHistory(
+    pipelineId: string,
+    maxQuantity?: number
+  ): Promise<PipelineHistoryEntry[]> {
+    const res = await this.db.query(
+      `
+    select * from podbox_edit_history
+    where pipeline->>'id' = $1
+    order by time_created asc
+    limit $2`,
+      [pipelineId, maxQuantity ?? 20]
+    );
+    return res.rows.map(
+      (row: PipelineHistoryRow) =>
+        ({
+          id: row.id,
+          pipeline: row.pipeline,
+          timeCreated: row.time_created.toISOString(),
+          editorUserId: row.editor_user_id
+        }) satisfies PipelineHistoryEntry
+    );
+  }
+}
+
+interface PipelineHistoryRow {
+  id: string;
+  pipeline: PipelineDefinition;
+  time_created: Date;
+  editor_user_id?: string;
```

Comment:
I think this type might be wrong - if the value in the DB is `NULL` then `editor_user_id` would be `null` and not undefined.

Reply:
you're right, fixed by updating the place a row gets converted into an object to coalesce both `null` and `undefined` to `undefined`